### PR TITLE
Bump minimum supported symengine version for built-in pickle support

### LINKED ISSUE #6188: Python 3.9 on linux hangs when using multiprocessing
### Information

- **Qiskit Terra version**: Master
- **Python version**: 3.9
- **Operating system**: Linux

### What is the current behavior?

On the linux 3.9 ci job there have been several jobs recently that fail due to a job timeout after 1 hour. For example:

https://dev.azure.com/qiskit-ci/qiskit-terra/_build/results?buildId=26619&view=logs&jobId=53f3e6ef-9cd2-51f8-4c23-ef405deeeb83&j=53f3e6ef-9cd2-51f8-4c23-ef405deeeb83&t=f9e6fb36-fdac-5369-6649-12f2d7a63b5d

Looking through the logs there, partway through the job test runner `{1}` stops reporting test status early on in the job. It is only partway through the `test.python.algorithms.test_vqe.TestVQE` test class when it stops reporting test status. This means one of the test methods it has yet to run in the class have likely gotten stuck for some reason and the job just hits the 1 hr timeout and exits.

### Steps to reproduce the problem

Run unit tests in the linux 3.9 ci environment.

### What is the expected behavior?

Test finish, either success or failure.

### Suggested solutions

Figure out the root cause and fix.

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

The new symengine 0.9 release added native support in the package for
pickling symengine objects. Previously we had been converting symengine
objects to sympy objects so we could pickle them. With native support
for pickle in symengine now we no longer need this which besides
removing unnecessary should hopefully make pickling (which we do
internally as part of using multiprocessing) more reliable.

This also seems to fix the hanging we were seeing with multiprocessing
with Python 3.9 on Linux. While investigating that issue it points to
the underlying cause being a bug in cPython with the `fork()` based
start method, but we were only able to reliably trigger it after
switching to symengine in #6270 and having to rely on importing
symengine to pickle the symengine objects. Since we're no longer doing
that after bumping the minimum symengine version this removes the
default disabling of parallel dispatch with Python 3.9. While I'm not
100% confident this fixes the bug, in my testing locally I haven't been
able to reproduce the hang we were encountering (but this is ancedotal
at best). If we do encounter issues with multiprocess hanging in the
future we can look at rewriting the internals of `parallel_map()` or
switching it back to disabled by default.

### Details and comments

Fixes #6188

### CODE REVIEW

File: releasenotes/notes/bump-symengine-8ca362f5b9fef199.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+---
+upgrade:
+  - |
+    The minimum supported version of ``symengine`` is now 0.9.0. This was
+    necessary to improve compatibility with Python's ``pickle`` module which
+    is used internally as part of parallel dispatch with :func:`.parallel_map`.
+  - |
+    The default value of ``QISKIT_PARALLEL` when running with Python 3.9 on
```
Comment (jakelishman) [inline releasenotes/notes/bump-symengine-8ca362f5b9fef199.yaml:8]:
```suggestion
    The default value of ``QISKIT_PARALLEL`` when running with Python 3.9 on
```

File: releasenotes/notes/bump-symengine-8ca362f5b9fef199.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    Linux is now set to ``TRUE``. This means when running :func:`.parallel_map`
+    or functions that call it internally such as :func:`.transpile` and
+    :func:`.assemble` the function will be executed in multiple processes and
+    should have better run time performance. This change was made because the
+    issues with reliability of parallel dispatch appear to have been resovled
```
Comment (jakelishman) [inline releasenotes/notes/bump-symengine-8ca362f5b9fef199.yaml:13]:
```suggestion
    issues with reliability of parallel dispatch appear to have been resolved
```

**Review by jakelishman — COMMENTED** [review]
This looks fine to me, and I'll be glad if it helps out with anything multi-processing related.

As an aside, do you have the benchmarks for how this impacts circuit construction with large numbers of `ParameterExpression`?  I'm hopeful we'll see a non-trivial improvement, because removing the `sympify` call should make `deepcopy` (somewhat) lighter.

**Review by jakelishman — APPROVED** [review]

**Comment by mtreinish** [comment]
> As an aside, do you have the benchmarks for how this impacts circuit construction with large numbers of `ParameterExpression`? I'm hopeful we'll see a non-trivial improvement, because removing the `sympify` call should make `deepcopy` (somewhat) lighter.

I haven't run benchmarks of this yet. I was more just happy to fix the multiprocessing issues. I'm doing benchmarking on #7658 right now, I'll spin up an asv run of the parameter expression benchmarks after I finish that (or we can just wait until tomorrow for the nightly run)

**Comment by jakelishman** [comment]
It can wait - it's more out of interest than anything else.

**Pull request merged**